smithy-diff/src/main/java/software/amazon/smithy/diff/evaluators/ChangedNullability.java

```diff
@@ -91,37 +93,60 @@ private ValidationEvent createError(
         boolean newHasInput = hasInputTrait(differences.getNewModel(), newMember);
+        Shape newTarget = differences.getNewModel().expectShape(newMember.getTarget());
+        Severity severity = null;
 
         if (oldHasInput && !newHasInput) {
             // If there was an input trait before, but not now, then the nullability must have
             // changed from nullable to non-nullable.
             joiner.add("The @input trait was removed from " + newMember.getContainer());
+            severity = Severity.ERROR;
         } else if (!oldHasInput && newHasInput) {
             // If there was no input trait before, but there is now, then the nullability must have
             // changed from non-nullable to nullable.
             joiner.add("The @input trait was added to " + newMember.getContainer());
+            severity = Severity.DANGER;
         } else if (!newHasInput) {
             // Can't add nullable to a preexisting required member.
             if (change.isTraitAdded(ClientOptionalTrait.ID) && change.isTraitInBoth(RequiredTrait.ID)) {
                 joiner.add("The @nullable trait was added to a @required member.");
+                severity = Severity.ERROR;
             }
             // Can't add required to a member unless the member is marked as nullable.
             if (change.isTraitAdded(RequiredTrait.ID) && !newMember.hasTrait(ClientOptionalTrait.ID)) {
                 joiner.add("The @required trait was added to a member that is not marked as @nullable.");
+                severity = Severity.ERROR;
             }
             // Can't add the default trait to a member unless the member was previously required.
             if (change.isTraitAdded(DefaultTrait.ID) && !change.isTraitRemoved(RequiredTrait.ID)) {
                 joiner.add("The @default trait was added to a member that was not previously @required.");
+                severity = Severity.ERROR;
             }
             // Can only remove the required trait if the member was nullable or replaced by the default trait.
             if (change.isTraitRemoved(RequiredTrait.ID)
                     && !newMember.hasTrait(DefaultTrait.ID)
                     && !oldMember.hasTrait(ClientOptionalTrait.ID)) {
-                joiner.add("The @required trait was removed and not replaced with the @default trait.");
+                if (newTarget.isStructureShape() || newTarget.isUnionShape()) {
+                    severity = severity == null ? Severity.WARNING : severity;
+                    joiner.add("The @required trait was removed from a member that targets a " + newTarget.getType()
+                               + ". This is backward compatible in generators that always treat structures and "
+                               + "unions as optional (e.g., AWS generators)");
+                } else {
+                    joiner.add("The @required trait was removed and not replaced with the @default trait.");
```

Comment:
Does the message also need `@addedDefault`?

```suggestion
                    joiner.add("The @required trait was removed and not replaced with the @default and @addedDefault traits.");
```